### DISCUSSION TITLE
Counting nodes in XML subtrees (with BIRT)

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [eclipse-birt/birt#906] [Q&A]
I specified a [computed column](https://eclipse.github.io/birt-website/docs/t_data-sets#data-set-editor-pages "Documentation") where entries should be counted according to a bit of test data.
I observed then that the item incidence was determined for all data records.

Query result preview:
![incidence-preview-example-20220329](https://user-images.githubusercontent.com/102525885/160626227-10e6338b-fd0d-43de-926a-8253d4a5f9c0.png)

:crystal_ball:
* How can it be achieved that the incidence would be calculated only for selected data groups (by the means of the report designer 4.9.0)?
* Can aggregate functions be applied also for XML data similar to a SQL clause like “`GROUP BY`”?

**Comment by hvbtup** [comment]
The correct place to do aggregation is in the layout, not in the dataset.

A table or list must be bound to a dataset.
For grouping and aggregation, you can add groups to `Table` and `List` items.
Each group of a list or a table has a header area and a footer area. For tables, these areas are rows.
The easiest way to add add aggregates is to drag the sigma symbol from the Quick Tools palette and then drop it into a header or footer of a group (or the whole list/table).
BIRT will then use the correct level automatically.

You can use many of the grouping functions you know from SQL.
It doesn't matter what kind of dataset the list/table is based on.

**Comment by Markus-Elfring** [comment]
>The correct place to do aggregation is in the layout, not in the dataset.

I got an other impression for my data processing requirements.
I find that I need to perform extra computations on demand also for my data sets.


>You can use many of the grouping functions you know from SQL.

[Some record sets are groupable by report layouts with tables](https://eclipse.github.io/birt-website/docs/t_tables#creating-groups-on-your-own "Creating groups on your own").
But I would like to extract information from hierarchical data structures (like XML) so that it becomes usable as additional values for each record set (according to the relational data model).
* I am going to work more with XQuery scripts.
* Is the support still evolving for the standard “SQL/XML”?

**Reply by hvbtup** [reply]
IIRC you can use an XML subtree as input to a second dataset (and so on) and create multi-level master-detail reports that way.
Perhaps someone else can shed some light on this?

**Reply by Markus-Elfring** [reply]
>you can use an XML subtree as input to a second dataset

This information sounds promising.
I imagine then that each XML subtree would need a parent context as a corresponding parameter so that the referential integrity can be preserved.


>and create multi-level master-detail reports that way.

I got interested in the generation of histograms for further data analyses.

**Comment by hvbtup** [comment]
This is not directly related to your initial question, but I created an example report that shows one way (probably there are others) how master-detail reports can be created with an XML data source.
[xml_quelle.zip](https://github.com/eclipse/birt/files/8381087/xml_quelle.zip)

To run the report, select an *.rptlibrary with a few different master pages as XML input for the data source `xmlds`.

The report will list the master pages in the library (master) and their properties (detail).

Note: BIRT is slow when processing big XML data sources (#762).
